scripts/init_model.py

```diff
@@ -0,0 +1,174 @@
+import tarfile
+import gzip
+import zipfile
+import json
+
+sys.path.insert(0, os.path.dirname(os.path.abspath(os.path.join(__file__, os.pardir))))
+
+from spacy.cli._messages import Messages
+from spacy.vectors import Vectors
+from spacy.errors import Errors, Warnings, user_warning
+from spacy.util import prints, ensure_path, get_lang_class
+
+from scispacy.file_cache import cached_path
+
+@plac.annotations(
+        lang=("model language", "positional", None, str),
+        output_dir=("model output directory", "positional", None, Path),
+        freqs_loc=("location of words frequencies file", "positional", None, Path),
+        vectors_loc=("optional: location of vectors file in Word2Vec format "
+                     "(either as .txt or zipped as .zip or .tar.gz)", "option",
+                     "v", str),
+        no_expand_vectors=("optional: Whether to expand vocab with words found in vector file",
+                           "flag", "x", bool),
+        meta_overrides=("optional: meta_json file to load.",
+                        "option", "m", Path),
+        prune_vectors=("optional: number of vectors to prune to",
+                       "option", "V", int)
+)
+def init_model(lang, output_dir, freqs_loc=None,
+               vectors_loc=None, no_expand_vectors=False,
+               meta_overrides=None, prune_vectors=-1):
+    """
+    Create a new model from raw data, like word frequencies, Brown clusters
+    and word vectors.
+    """
+    if vectors_loc is not None:
+        vectors_loc = cached_path(vectors_loc)
+    freqs_loc = cached_path(freqs_loc)
+
+    if freqs_loc is not None and not freqs_loc.exists():
+        prints(freqs_loc, title=Messages.M037, exits=1)
+    vectors_loc = ensure_path(vectors_loc)
+    probs, oov_prob = read_freqs(freqs_loc) if freqs_loc is not None else ({}, -20)
+    vectors_data, vector_keys = read_vectors(vectors_loc) if vectors_loc else (None, None)
+    nlp = create_model(lang, probs, oov_prob, vectors_data, vector_keys, not no_expand_vectors, prune_vectors)
+
+    if meta_overrides is not None:
+        metadata = json.load(open(meta_overrides))
+        nlp.meta.update(metadata)
+
+    if not output_dir.exists():
+        output_dir.mkdir()
+    nlp.to_disk(output_dir)
+    return nlp
+
+def open_file(loc):
+    '''Handle .gz, .tar.gz or unzipped files'''
+    loc = ensure_path(loc)
+    print("Open loc")
+    if tarfile.is_tarfile(str(loc)):
+        return tarfile.open(str(loc), 'r:gz')
+    elif loc.parts[-1].endswith('gz'):
+        return (line.decode('utf8') for line in gzip.open(str(loc), 'r'))
+    elif loc.parts[-1].endswith('zip'):
+        zip_file = zipfile.ZipFile(str(loc))
+        names = zip_file.namelist()
+        file_ = zip_file.open(names[0])
+        return (line.decode('utf8') for line in file_)
+    else:
+        return loc.open('r', encoding='utf8')
+
+def create_model(lang, probs, oov_prob, vectors_data, vector_keys, expand_vectors, prune_vectors):
```

Comment:
If I'm reading this correctly, this means that if this function is called with vectors/probs empty/None, as it could be based on optional arguments to init_model, that lang class returned won't have a vocab. Should this function default to using the default vocab?

Reply:
The plac annotations don't actually specify default values for positional arguments (i.e it crashes) I agree this is very confusing. I was mostly trying to stay close to the spacy script as i'm going to submit these additions upstream, so hopefully eventually we can use them and remove this script